tests/integration/telemetry/stats/prometheus/wasm/imagepullpolicy_test.go

```diff
@@ -0,0 +1,251 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/scopes"
+	"istio.io/istio/pkg/test/util/retry"
+	common "istio.io/istio/tests/integration/telemetry/stats/prometheus"
+)
+
+const (
+	imageName      = "istio-testing/wasm/header-injector"
+	injectedHeader = "x-resp-injection"
+	wasmConfigFile = "testdata/wasm-filter.yaml"
+)
+
+type wasmTestConfigs struct {
+	desc            string
+	name            string
+	policy          string
+	tag             string
+	upstreamVersion string
+	expectedVersion string
+}
+
+var generation = 0
+
+func mapTagToVersionOrFail(t framework.TestContext, tag, version string) {
+	t.Helper()
+	if err := registry.SetupTagMap(map[string]string{
+		imageName + ":" + tag: version,
+	}); err != nil {
+		t.Fatalf("failed to setup the tag map: %v", err)
+	}
+}
+
+func applyAndTestWasm(t framework.TestContext, c wasmTestConfigs) {
+	t.Helper()
+	defer func() {
+		generation++
+	}()
+	scopes.Framework.Infof(c.desc)
+	mapTagToVersionOrFail(t, c.tag, c.upstreamVersion)
+	if err := installWasmExtension(t, c.name, c.tag, c.policy, fmt.Sprintf("g-%q", generation)); err != nil {
+		t.Fatalf("failed to install WasmPlugin: %v", err)
+	}
+	retry.UntilSuccessOrFail(t, func() error {
+		result, err := sendTraffic()
+		if err != nil {
+			t.Log("failed to send traffic")
+			return err
+		}
+
+		if result.Responses.Len() == 0 {
+			return errors.New("failed to get response")
+		}
+
+		value := result.Responses[0].ResponseHeaders.Get(injectedHeader)
+		if value != c.expectedVersion {
+			return fmt.Errorf("[generation: %d] unexpected values for the header %q with the policy %q: given %q, got %q, want %q",
+				generation, injectedHeader, c.policy, c.upstreamVersion, value, c.expectedVersion)
+		}
+
+		return nil
+	}, retry.Delay(1*time.Second), retry.Timeout(100*time.Second))
+}
+
+func resetWasm(t framework.TestContext, pluginName string) {
+	t.Helper()
+	scopes.Framework.Infof("Delete WasmPlugin %q", pluginName)
+	if err := uninstallWasmExtension(t, pluginName); err != nil {
+		t.Fatal(err)
+	}
+
+	successCount := 0
+
+	retry.UntilSuccessOrFail(t, func() error {
+		result, err := sendTraffic()
+		if err != nil {
+			t.Log("failed to send traffic")
+			return err
+		}
+
+		if result.Responses.Len() == 0 {
+			return errors.New("failed to get response")
+		}
+
+		value := result.Responses[0].ResponseHeaders.Get(injectedHeader)
+
+		if value != "" {
+			return fmt.Errorf("failed to reset WasmPlugin %q", pluginName)
+		}
+
+		if successCount < 2 {
```

Comment:
can pass retry.Convege(2) to CallOptions.Retry.Options

Reply:
Done!